File: plsql-frontend/src/main/java/org/sonar/plsqlopen/squid/ProgressReport.java
/*
 * Sonar PL/SQL Plugin (Community)
 * Copyright (C) 2015-2018 Felipe Zorzo
 * mailto:felipebzorzo AT gmail DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plsqlopen.squid;

import java.util.Collection;
import java.util.Iterator;

import org.sonar.api.batch.fs.InputFile;
import org.sonar.api.utils.log.Logger;
import org.sonar.api.utils.log.Loggers;

public class ProgressReport implements Runnable {

    private final long period;
    private final Logger logger;
    private int count;
    private int currentFileNumber = -1;
    private InputFile currentFile;
    private Iterator<InputFile> it;
    private final Thread thread;
    private final String adjective;
    private boolean success = false;

    public ProgressReport(String threadName, long period, Logger logger, String adjective) {
        this.period = period;
        this.logger = logger;
        this.adjective = adjective;
        thread = new Thread(this);
        thread.setName(threadName);
        thread.setDaemon(true);
    }

    public ProgressReport(String threadName, long period, String adjective) {
        this(threadName, period, Loggers.get(ProgressReport.class), adjective);
    }

    public ProgressReport(String threadName, long period) {
        this(threadName, period, "analyzed");
    }

    @Override
    public void run() {
        while (!Thread.interrupted()) {
            try {
                Thread.sleep(period);
                synchronized (this) {
                    log(currentFileNumber + "/" + count + " files " + adjective + ", current file: "
                            + currentFile.toString());
                }
            } catch (InterruptedException e) {
                break;
            }
        }
        synchronized (this) {
            if (success) {
                log(count + "/" + count + " source files have been " + adjective);
            }
        }
    }

    public synchronized void start(Collection<InputFile> files) {
        count = files.size();
        it = files.iterator();

        nextFile();

        log(count + " source files to be " + adjective);
        thread.start();
    }

    public synchronized void nextFile() {
        if (it.hasNext()) {
            currentFileNumber++;
            currentFile = it.next();
        }
    }

    public synchronized void stop() {
        success = true;
        thread.interrupt();
    }

    public synchronized void cancel() {
        thread.interrupt();
    }

    public void join() throws InterruptedException {
        thread.join();
    }

    private void log(String message) {
        synchronized (logger) {
            logger.info(message);
            logger.notifyAll();
        }
    }

}

File: zpa-core/src/main/kotlin/org/sonar/plsqlopen/squid/ProgressReport.kt
/**
 * Z PL/SQL Analyzer
 * Copyright (C) 2015-2020 Felipe Zorzo
 * mailto:felipebzorzo AT gmail DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plsqlopen.squid

import org.sonar.plsqlopen.utils.log.Logger
import org.sonar.plsqlopen.utils.log.Loggers
import java.util.concurrent.locks.Condition
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class ProgressReport @JvmOverloads constructor(threadName: String,
                                               private val period: Long,
                                               private val logger: Logger = Loggers.getLogger(ProgressReport::class.java)) :
    Runnable {

    internal val lock = ReentrantLock()
    internal val condition: Condition = lock.newCondition()

    private var count: Int = 0
    private var currentFileNumber = -1
    private var currentFile: String? = null
    private lateinit var iterator: Iterator<String>
    private val thread: Thread = Thread(this)
    private var success = false

    init {
        thread.name = threadName
        thread.isDaemon = true
    }

    override fun run() {
        while (!Thread.interrupted()) {
            try {
                Thread.sleep(period)
                lock.withLock {
                    log("$currentFileNumber/$count files analyzed, current file: $currentFile")
                }
            } catch (e: InterruptedException) {
                break
            }

        }

        lock.withLock {
            if (success) {
                log("$count/$count source files have been analyzed")
            }
        }
    }

    @Synchronized
    fun start(files: Collection<String>) {
        count = files.size
        iterator = files.iterator()

        nextFile()

        log("$count source files to be analyzed")
        thread.start()
    }

    @Synchronized
    fun nextFile() {
        if (iterator.hasNext()) {
            currentFileNumber++
            currentFile = iterator.next()
        }
    }

    @Synchronized
    fun stop() {
        success = true
        thread.interrupt()
    }

    @Synchronized
    fun cancel() {
        thread.interrupt()
    }

    @Throws(InterruptedException::class)
    fun join() {
        thread.join()
    }

    private fun log(message: String) {
        lock.withLock {
            logger.info(message)
            condition.signalAll()
        }
    }

}
